File: src/simulation/EventGenerator.cs
using ElevenLegends.Data.Enums;
using ElevenLegends.Data.Models;

namespace ElevenLegends.Simulation;

/// <summary>
/// Generates match events from action results.
/// </summary>
public static class EventGenerator
{
    /// <summary>
    /// Generates zero or more events from an action result.
    /// </summary>
    public static List<MatchEvent> Generate(ActionResult result, int tick, Player? assistProvider = null)
    {
        var events = new List<MatchEvent>();

        if (result.IsGoal)
        {
            events.Add(new MatchEvent
            {
                Tick = tick,
                Type = EventType.Goal,
                PlayerId = result.Executor.Id,
                SecondaryPlayerId = assistProvider?.Id,
                Description = $"GOAL! {result.Executor.Name} scores!",
                RatingImpact = 1.5f
            });

            if (assistProvider != null)
            {
                events.Add(new MatchEvent
                {
                    Tick = tick,
                    Type = EventType.Assist,
                    PlayerId = assistProvider.Id,
                    Description = $"Assist by {assistProvider.Name}",
                    RatingImpact = 1.0f
                });
            }
        }
        else if (result.IsShotOnTarget)
        {
            events.Add(new MatchEvent
            {
                Tick = tick,
                Type = EventType.ShotOnTarget,
                PlayerId = result.Executor.Id,
                Description = $"Shot on target by {result.Executor.Name} — saved!",
                RatingImpact = 0.1f
            });

            // GK save event would be handled separately
        }
        else if (result.Action == ActionType.Shot && !result.Success)
        {
            events.Add(new MatchEvent
            {
                Tick = tick,
                Type = EventType.Shot,
                PlayerId = result.Executor.Id,
                Description = $"Shot off target by {result.Executor.Name}",
                RatingImpact = -0.1f
            });
        }

        if (result.IsFoul)
        {
            events.Add(new MatchEvent
            {
                Tick = tick,
                Type = EventType.Foul,
                PlayerId = result.Executor.Id,
                SecondaryPlayerId = result.Opponent?.Id,
                Description = $"Foul by {result.Executor.Name}",
                RatingImpact = -0.2f
            });
        }

        // Successful defensive actions
        if (result.Success && result.Action == ActionType.Tackle && !result.IsFoul)
        {
            events.Add(new MatchEvent
            {
                Tick = tick,
                Type = EventType.Save, // reusing as "defensive action" for now
                PlayerId = result.Executor.Id,
                Description = $"Great tackle by {result.Executor.Name}",
                RatingImpact = 0.3f
            });
        }

        if (result.Success && result.Action == ActionType.Interception)
        {
            events.Add(new MatchEvent
            {
                Tick = tick,
                Type = EventType.Save,
                PlayerId = result.Executor.Id,
                Description = $"Interception by {result.Executor.Name}",
                RatingImpact = 0.3f
            });
        }

        // Successful dribble
        if (result.Success && result.Action == ActionType.Dribble)
        {
            events.Add(new MatchEvent
            {
                Tick = tick,
                Type = EventType.Shot, // generic positive action event
                PlayerId = result.Executor.Id,
                Description = $"Successful dribble by {result.Executor.Name}",
                RatingImpact = 0.2f
            });
        }

        // Failed pass
        if (!result.Success && result.Action == ActionType.Pass)
        {
            events.Add(new MatchEvent
            {
                Tick = tick,
                Type = EventType.Shot, // generic event
                PlayerId = result.Executor.Id,
                Description = $"Bad pass by {result.Executor.Name}",
                RatingImpact = -0.2f
            });
        }

        // Turnover on failed dribble
        if (!result.Success && result.Action == ActionType.Dribble)
        {
            events.Add(new MatchEvent
            {
                Tick = tick,
                Type = EventType.Shot,
                PlayerId = result.Executor.Id,
                Description = $"Lost possession: {result.Executor.Name}",
                RatingImpact = -0.3f
            });
        }

        return events;
    }
}

File: engine/simulation/event-generator.ts
import { ActionType, EventType } from "@/engine/enums";
import type { MatchEvent, Player } from "@/engine/models";
import type { ActionResult } from "./action-resolver";

/**
 * Generates match events from action results.
 */

/** Generates zero or more events from an action result. */
export function generate(
  result: ActionResult,
  tick: number,
  assistProvider?: Player | null,
): MatchEvent[] {
  const events: MatchEvent[] = [];

  if (result.isGoal) {
    events.push({
      tick,
      type: EventType.Goal,
      playerId: result.executor.id,
      secondaryPlayerId: assistProvider?.id ?? null,
      description: `GOAL! ${result.executor.name} scores!`,
      ratingImpact: 1.5,
    });

    if (assistProvider != null) {
      events.push({
        tick,
        type: EventType.Assist,
        playerId: assistProvider.id,
        secondaryPlayerId: null,
        description: `Assist by ${assistProvider.name}`,
        ratingImpact: 1.0,
      });
    }
  } else if (result.isShotOnTarget) {
    events.push({
      tick,
      type: EventType.ShotOnTarget,
      playerId: result.executor.id,
      secondaryPlayerId: null,
      description: `Shot on target by ${result.executor.name} -- saved!`,
      ratingImpact: 0.1,
    });

    // GK save event would be handled separately
  } else if (result.action === ActionType.Shot && !result.success) {
    events.push({
      tick,
      type: EventType.Shot,
      playerId: result.executor.id,
      secondaryPlayerId: null,
      description: `Shot off target by ${result.executor.name}`,
      ratingImpact: -0.1,
    });
  }

  if (result.isFoul) {
    events.push({
      tick,
      type: EventType.Foul,
      playerId: result.executor.id,
      secondaryPlayerId: result.opponent?.id ?? null,
      description: `Foul by ${result.executor.name}`,
      ratingImpact: -0.2,
    });
  }

  // Successful defensive actions
  if (
    result.success &&
    result.action === ActionType.Tackle &&
    !result.isFoul
  ) {
    events.push({
      tick,
      type: EventType.Save, // reusing as "defensive action" for now
      playerId: result.executor.id,
      secondaryPlayerId: null,
      description: `Great tackle by ${result.executor.name}`,
      ratingImpact: 0.3,
    });
  }

  if (result.success && result.action === ActionType.Interception) {
    events.push({
      tick,
      type: EventType.Save,
      playerId: result.executor.id,
      secondaryPlayerId: null,
      description: `Interception by ${result.executor.name}`,
      ratingImpact: 0.3,
    });
  }

  // Successful dribble
  if (result.success && result.action === ActionType.Dribble) {
    events.push({
      tick,
      type: EventType.Shot, // generic positive action event
      playerId: result.executor.id,
      secondaryPlayerId: null,
      description: `Successful dribble by ${result.executor.name}`,
      ratingImpact: 0.2,
    });
  }

  // Failed pass
  if (!result.success && result.action === ActionType.Pass) {
    events.push({
      tick,
      type: EventType.Shot, // generic event
      playerId: result.executor.id,
      secondaryPlayerId: null,
      description: `Bad pass by ${result.executor.name}`,
      ratingImpact: -0.2,
    });
  }

  // Turnover on failed dribble
  if (!result.success && result.action === ActionType.Dribble) {
    events.push({
      tick,
      type: EventType.Shot,
      playerId: result.executor.id,
      secondaryPlayerId: null,
      description: `Lost possession: ${result.executor.name}`,
      ratingImpact: -0.3,
    });
  }

  return events;
}
